library/std/src/sys/windows/compat.rs

```diff
@@ -28,45 +27,53 @@ pub fn lookup(module: &str, symbol: &str) -> Option<usize> {
     }
 }
 
-pub fn store_func(ptr: &AtomicUsize, module: &str, symbol: &str, fallback: usize) -> usize {
-    let value = lookup(module, symbol).unwrap_or(fallback);
-    ptr.store(value, Ordering::SeqCst);
-    value
-}
-
 macro_rules! compat_fn {
-    ($module:ident: $(
+    ($module:literal: $(
         $(#[$meta:meta])*
-        pub fn $symbol:ident($($argname:ident: $argtype:ty),*)
-                                  -> $rettype:ty {
-            $($body:expr);*
-        }
+        pub fn $symbol:ident($($argname:ident: $argtype:ty),*) -> $rettype:ty $body:block
     )*) => ($(
-        #[allow(unused_variables)]
         $(#[$meta])*
-        pub unsafe fn $symbol($($argname: $argtype),*) -> $rettype {
+        pub mod $symbol {
+            use super::*;
             use crate::sync::atomic::{AtomicUsize, Ordering};
             use crate::mem;
-            type F = unsafe extern "system" fn($($argtype),*) -> $rettype;
 
             static PTR: AtomicUsize = AtomicUsize::new(0);
 
+            #[allow(unused_variables)]
+            unsafe extern "system" fn fallback($($argname: $argtype),*) -> $rettype $body
+
+            #[cold]
             fn load() -> usize {
-                crate::sys::compat::store_func(&PTR,
-                                          stringify!($module),
-                                          stringify!($symbol),
-                                          fallback as usize)
+                // There is no locking here. It's okay if this is executed by multiple threads in
+                // parallel. `lookup` will result in the same value, and it's okay if they overwrite
+                // eachothers result as long as they do so atomically. We don't need any guarantees
+                // about memory ordering, as this involves just a single atomic variable which is
+                // not used to protect or order anything else.
+                let addr = crate::sys::compat::lookup($module, stringify!($symbol))
+                    .unwrap_or(fallback as usize);
+                PTR.store(addr, Ordering::Relaxed);
+                addr
             }
-            unsafe extern "system" fn fallback($($argname: $argtype),*)
-                                               -> $rettype {
-                $($body);*
+
+            fn addr() -> usize {
+                match PTR.load(Ordering::Relaxed) {
+                    0 => load(),
+                    addr => addr,
+                }
             }
 
-            let addr = match PTR.load(Ordering::SeqCst) {
-                0 => load(),
-                n => n,
-            };
-            mem::transmute::<usize, F>(addr)($($argname),*)
+            #[allow(dead_code)]
+            pub fn is_available() -> bool {
+                addr() != fallback as usize
```

Comment:
Function pointer equality is not well defined, since we give functions to LLVM as `unnamed_addr` which means their address is claimed to not be significant to the execution of the program. See https://github.com/rust-lang/rust/issues/54685. This expression can evaluate to unequal even if lookup failed, and can evaluate to equal even if lookup succeeded. Is there a different way to implement is_available, or is there some reason to believe this wouldn't hit bizarre behavior?

Reply:
Oh, interesting. Thanks.

If I understand correctly, `unnamed_addr` allows for merging functions. Meaning that `f1 == f2` might give true for different functions, but not that `f1 == f1` might give false for the same function. So false positives, but not false negatives. That would be fine here. The functions loaded from `GetProcAddress` are definitely not merged with `fallback` (since they come from external `dll`), and that's the only one that matters. If multiple `fallback` functions happen to have the same address for example, that's fine.

So the question here is whether 'the address is not significant' means 'not significant but at least constant', or that it may assume`f != f`.

Disabling `unnamed_addr` on `fallback` would also work, but it seems there's no way to do that in Rust.

Reply:
Ah, @eddyb tells me *rustc itself* might *duplicate* functions too. So, in this snippet it could mean that `fallback as usize` (which is short for `fallback as fn() as usize`) might result in a copy of `fallback`. Using a `static FALLBACK: usize = fallback as usize;` and using that static instead should prevent that. The assumption that that works is already made in `core`:
https://github.com/rust-lang/rust/blob/9cba260df0f1c67ea3690035cd5611a7465a1560/library/core/src/fmt/mod.rs#L246-L267

I'll update the PR to do the same.